Libplanet/Net/NetMQTransport.cs

```diff
@@ -0,0 +1,1009 @@
+            }
+
+            _router = new RouterSocket();
+            _router.Options.RouterHandover = true;
+
+            if (_host is null && !(_iceServers is null))
+            {
+                _turnClient = await IceServer.CreateTurnClient(_iceServers);
+            }
+
+            if (_listenPort == null)
+            {
+                _listenPort = _router.BindRandomPort("tcp://*");
+            }
+            else
+            {
+                _router.Bind($"tcp://*:{_listenPort}");
+            }
+
+            _logger.Information($"Listen on {_listenPort}");
+
+            _cancellationToken = cancellationToken;
+            var tasks = new List<Task>();
+
+            if (!(_turnClient is null))
+            {
+                _publicIPAddress = (await _turnClient.GetMappedAddressAsync()).Address;
+
+                if (await _turnClient.IsBehindNAT())
+                {
+                    _behindNAT = true;
+                }
+            }
+
+            if (_behindNAT)
+            {
+                IPEndPoint turnEp = await _turnClient.AllocateRequestAsync(
+                    TurnAllocationLifetime
+                );
+                EndPoint = new DnsEndPoint(turnEp.Address.ToString(), turnEp.Port);
+
+                // FIXME should be parameterized
+                tasks.Add(BindingProxies(_cancellationToken));
+                tasks.Add(BindingProxies(_cancellationToken));
+                tasks.Add(BindingProxies(_cancellationToken));
+                tasks.Add(RefreshAllocate(_cancellationToken));
+                tasks.Add(RefreshPermissions(_cancellationToken));
+            }
+            else if (_host is null)
+            {
+                EndPoint = new DnsEndPoint(_publicIPAddress.ToString(), _listenPort.Value);
+            }
+            else
+            {
+                EndPoint = new DnsEndPoint(_host, _listenPort.Value);
+            }
+
+            _replyQueue = new NetMQQueue<Message>();
+            _broadcastQueue = new NetMQQueue<(Address?, Message)>();
+            _poller = new NetMQPoller { _router, _replyQueue, _broadcastQueue };
+
+            _router.ReceiveReady += ReceiveMessage;
+            _replyQueue.ReceiveReady += DoReply;
+            _broadcastQueue.ReceiveReady += DoBroadcast;
+        }
+
+        public async Task RunAsync(CancellationToken cancellationToken)
+        {
+            if (Running)
+            {
+                throw new SwarmException("Swarm is already running.");
+            }
+
+            Running = true;
+
+            List<Task> tasks = new List<Task>();
+
+            tasks.Add(Task.Run(
+                () =>
+                {
+                    // Ignore NetMQ related exceptions during NetMQPoller.Run() to stabilize
+                    // tests.
+                    try
+                    {
+                        _poller.Run();
+                    }
+                    catch (TerminatingException)
+                    {
```

Comment:
It seems some lines needed more. see also [Swarm.cs#L514](https://github.com/planetarium/libplanet/blob/master/Libplanet/Net/Swarm.cs#L514).

Reply:
I added it :)